### ISSUE TITLE
Json Serialization Updates for Dynamo

### ISSUE THREAD

**Issue by ikeough** [DynamoDS/Dynamo#7747]
The purpose of this proposal is to define the json graph format for Dynamo graphs. This graph format splits the graph into two parts. The first part represents the topology of the graph using nodes and connectors. The second part contains objects related to the visualization of the graph. Everything not related to describing compute connectivity, or visualizing the graph, will be considered an application concern and will need to be stored elsewhere. An example of this would be `Nodes` versus `Cameras`. `Nodes` are required in the graph to describe function calls, but `Cameras` are only used by current Dynamo applications to save vector information related to the user's viewpoint when visualizing the scene in 3D.

### Motivations
- One graph definition shared between Dynamo, CoGS/TS, and AVP.
- The ability to add custom converters to serialize to property sets.
- Direct serialization of `NodeModels`. No more `Serialize()` and `Deserialize()` methods. We will be able to remove inconsistent xml serialization code. Node authors have used attributes and elements non-discriminantly over the years, resulting in bloated parsing and deserialization code. 
- Refactoring of Dynamo base classes. Properties related to visualization of the graph should live in the view model or the view.
- Move versioning from Dynamo to the graph. As more systems rely on the shared graph format, it will be necessary to have a version which means something across the entire system. It is possible that this version can also be tied to the design script language, and that bi-directional conversion between DS and the graph is a requirement.

### What Has Already Been Done
- The graph schema for CoGS. https://dev-cogs.thunderstorm.autodesk.com/computation-graph/v1/docs#operation/executeGraph
- Attribution of Existing Classes to Support Serialization to Json. https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Graph/Nodes/NodeModel.cs), although it is not used as Dynamo's file format.
- Command line support for graph conversion to json.

### Additional Questions and Considerations
- Does the json graph format become Dynamo's _one_ graph format? If so, we will need to edit the proposal below to consider custom nodes.
- If `Port` descriptions are not stored in the graph, where are they stored? Is it possible to have shared function meta-data across Dynamo, AVP, CoGS, and TS.
- One source of truth for port and function descriptions. This would benefit localization.

### The Proposed Graph Format
```json
{
  "Id": "f170f3ea-ae94-43c1-af0c-e2fa76ebc725",
  "Name":"My graph",
  "Description":"A compute graph.",
  "Nodes":[
    {
      "Id":"aNode",
      "ConcreteType":"Dynamo.Graph.Nodes.ZeroTouch.DSFunction, DynamoCore",
      "NodeType":"FunctionNode",
      "FunctionSignature":"Autodesk.DesignScript.Geometry.Point.ByCoordinates@double,double,double",
      "VariableInputPorts": true,
      "Inputs":[
        {
          "Id":"aPort",
          "UseLevels":false,
          "SelectedLevel":2,
          "KeepListStructure":false
        }
      ],
      "Outputs":[],
      "Replication":"auto"
    }
  ],
  "Connectors":[
    {
      "Id": "593c497e-08e3-485a-bf5d-1943d87a9c3b",
      "Start": "38727d94-86bd-4db8-b2a3-95baa1ac8a85",
      "End": "54c0f71a-524b-4443-9972-1e433a0dceb4"
    }
  ],
  "Dependencies":[
    "b7d9beae-14a0-485b-b259-27d0861eb165",
    "e1764a07-f1f4-4257-a672-9920cd7acc4a"
  ],
  "Bindings":[],
  "View":{
    "X":0.0,
    "Y":1000.0,
    "Zoom":0.5,
    "NodeViews":[
      {
        "d65016b6-6845-4423-a831-d20cbf7d37d4":{
          "Name":"Point.ByCoordinates",
          "X":0.0,
          "Y":5.0,
          "IsVisible":"true",
          "IsUpstreamVisible":true
        }
      }
    ],
    "Cameras":[],
    "Annotations":[
      {
        "Id": "A group of nodes.",
        "Nodes": [
          "61aa6da7-0f40-4fbd-b600-05e0daf7e70c",
          "6f0d0008-b446-40aa-be41-75fb3b375f39",
          "00fdf5cf-f509-40d0-afbb-e83957b19cf3",
          "3ea93ee6-7419-4e6a-9748-25737de09ff4"
        ]
      }
    ]
  }
}
```
### Proposed Changes

New JSON Graph Format | Current JSON Graph Format | Updates to Dynamo | Comments |
---|---|---|---|
**Graph** |
`Uuid` | `Uuid` | `Id` | The unique identifier of the graph.
`Name`| `Name` | | The user-defined name of the graph.
 | `LastModified` | Remove from serialization. | The date/time representing the last update of the graph.
 | `LastModifiedBy` | Remove from serialization. | The unique id of the last user who modified the graph.
`Description` | `Description` | | A description of the graph.
`Nodes` | `Nodes` | | An array of `Node` objects.
`Connectors` | `Connectors` | | An array of `Connector` objects.
`Dependencies` | | Add a `Dependencies field to `WorkspaceModel`. | An array of identifiers of graphs that are used as functions in this graph.
`Bindings` | `Bindings` | Change `sessionData` to `Bindings`. | An array of base-64 encoded key value pairs which relate call sites in the graph to data in an associated data store
`View` | | Adjust serialization logic to optionally serialize workspace view data. | An _optional_ object which has properties related to the visualization state of the graph
**View** | 
`X` | `X` | Move to `WorkspaceViewModel`. | The x coordinate of the upper left corner of the graph relative to the current viewing frame.
`Y` | `Y` | Move to `WorkspaceViewModel`. | The Y coordinate of the upper left corner of the graph relative to the current viewing frame.
`Zoom` | `Zoom` | Move to `WorkspaceViewModel`. | The scale factor between 0.0 and 1.0 of the graph.
`Cameras` | `Cameras` | Move to `WorkspaceViewModel`. | An array of `Camera` objects.
`Annotations` | `Annotations` | Move to `WorkspaceViewModel`. | An array of `Annotation` objects.
`NodeViews` | | Adjust serialization logic to optionally serialize node view data. | A map of objects, keyed by the node id, which contains coordinates of each node.
**NodeView**
`Name` | `DisplayName` | Change to `Name`.
`X` | `X` | Move to `NodeViewModel`.
`Y` | `Y` | Move to `NodeViewModel`.
`IsVisible` | `IsVisible` | Move to `NodeViewModel`.
`IsUpstreamVisible` | `IsUpstreamVisible` | Move to `NodeViewModel`
**Port**
`Id` | `Uuid` | | The identifier of the port. Id should be unique within the graph.
`UseLevels` | `UseLevels` | | A bool which specifies whether the `selectedLevel` and `keepListStructure` properties are considered.
`Level` | `Level` | | The level, starting from 0 at the leaf nodes, of nested data which is to be extracted at this port.
`KeepListStructure` | `ShouldKeepListStructure` | Rename to `KeepListStructure`. | `KeepListStructure` is a bool which specifies whether the incoming list topology should be maintained in the final data.
**Node**
`Id` | `Id` | | The identifier of the graph. Id should be unique within the graph.
`ConcreteType` | `ConcreteType` | | The location of the function identified by the `FunctionSignature` property. For .net assemblies used in Dynamo, this will be a fully qualified type name like `Dynamo.Graph.Nodes.ZeroTouch.DSFunction, DynamoCore`.
`NodeType` | `NodeType` | | The node type to be used on the server.
`FunctionSignature` | `FunctionSignature` | | The full qualified name of a function, including its parameters. Parameters are required to disambiguate overloaded functions.
`Inputs` | `InputPorts` | Rename to `Inputs`. |  An array of `port` objects.
`Outputs` | `OutputPorts` | Rename to `Outputs`. |  An array of `port` objects. 
`VariableInputPorts` | `VariableInputPorts` |  | A bool which indicates whether the node has a variable number of input ports.
`Replication` | `Replication` |  | Used to identify the replication strategy used by this node.
**Annotation**
`Id` | `Uuid` | |  The identifier of that annotation. `Id` should be unique within the document.
`Nodes` | `SelectedModels` | Rename to `Nodes`. | An array of unique identifiers of nodes which are referenced by this annotation.

**Comment by ikeough** [comment]
@kronz I'd like your feedback on the `viewData` section idea. This separates the graph description required for compute and that for visualization.

**Comment by gregmarr** [comment]
We should use the names from the CoGS schema, so they should have initial caps.

**Comment by radumg** [comment]
Really like this idea - fits in well with all the web-centred work being done on Forge, and with the where we want to take things. I have 2 quick questions : 

### Bindings
Not sure if I understood the `bindings` one correctly. 
```bindings is an array of base-64 encoded key value pairs which relate call sites in the graph to data in an associated data store. This is currently called callSiteData.```

Is this data that Dynamo will use as inputs for specific nodes ? If yes, great!
 
If not, it would be good to add such a property to the JSON : the use case I have in mind is a **REST** service for running Dynamo graphs where you can POST to it everything it needs to run : a `.dyn` file and any associated `data` or `input values`, returning the result (values, JSON or updated data files).

### YAML / RAML / Swagger
It would be great to publish & develop the API definitions in a common & purpose-made format so people can use them in building integrations and apps that interact with Dynamo.

**Comment by gregmarr** [comment]
@radumg That's not what that particular field is for.  Can you contact me offline?  first.last@autodesk.com

**Comment by ikeough** [comment]
@radumg We are building some services internally which will use this graph format. Those services are built using Swagger, so this graph format is actually represented as a JSON schema in the swagger. I'm not very happy with the client code generated by `swagger-codegen` currently, so for the immediate future, the .net clients (at least) for these services will be written by hand. For external users, they'll still be able to call `/ourShinyNewService/swagger.json` to get the schema and build clients of their own.

**Comment by pboyer** [comment]
I do not see why we are storing `lastModified` or `lastModifiedBy` data in the file format. This is not, strictly speaking, the responsibility of the graph format. On a desktop it would be handled by the file system. On the web, it really depends on the storage system, so it shouldn't be stored here.

**Comment by pboyer** [comment]
Appending the word `Data` onto a field of a data serialization format is a tautology. It could be removed with minimal loss in readability.

**Comment by pboyer** [comment]
Do all of these have to be a `uuid`? I'm guessing this is largely for historical purposes, but correct me if I'm wrong. Not all systems will generate equally good uuid's, some implementations are hacks or one-off jobs, so you may want to conceive the format in such a way that the responsibilities of the author and the system are explicit.

**Comment by ikeough** [comment]
@boyerp Agreed on the "viewData". We could just have "graph" and "view".

As for the uuids, we should make it clear that we mean this: https://www.iso.org/standard/53416.html

**Comment by ikeough** [comment]
We should also provide the required date/time format if we decide to keep any fields which store date/time. I know @marrg just ran into this with Dynamo's date/time node supporting any date/time format. This needs to be consistent.

**Comment by QilongTang** [comment]
@ikeough Green team discussed the json format today and thinking about the following format. The only difference between this and your proposed format is wrapping everything other than `view` blob into a `graph` blob. Let us know if you think it's OK. Also tasks are coming :)
```json
{
  "graph":{
    "Id": "f170f3ea-ae94-43c1-af0c-e2fa76ebc725",
    "Name":"My graph",
    "Description":"A compute graph.",
    "Nodes":[
      {
        "Id":"aNode",
        "ConcreteType":"Dynamo.Graph.Nodes.ZeroTouch.DSFunction, DynamoCore",
        "NodeType":"FunctionNode",
        "FunctionSignature":"Autodesk.DesignScript.Geometry.Point.ByCoordinates@double,double,double",
        "VariableInputPorts": true,
        "Inputs":[
          {
            "Id":"aPort",
            "UseLevels":false,
            "SelectedLevel":2,
            "KeepListStructure":false
          }
        ],
        "Outputs":[],
        "Replication":"auto"
      }
    ],
    "Connectors":[
      {
        "Id": "593c497e-08e3-485a-bf5d-1943d87a9c3b",
        "Start": "38727d94-86bd-4db8-b2a3-95baa1ac8a85",
        "End": "54c0f71a-524b-4443-9972-1e433a0dceb4"
      }
    ],
    "Dependencies":[
      "b7d9beae-14a0-485b-b259-27d0861eb165",
      "e1764a07-f1f4-4257-a672-9920cd7acc4a"
    ],
    "Bindings":[]
  }
  "View":{
    "X":0.0,
    "Y":1000.0,
    "Zoom":0.5,
    "NodeViews":[
      {
        "d65016b6-6845-4423-a831-d20cbf7d37d4":{
          "Name":"Point.ByCoordinates",
          "X":0.0,
          "Y":5.0,
          "IsVisible":"true",
          "IsUpstreamVisible":true
        }
      }
    ],
    "Cameras":[],
    "Annotations":[
      {
        "Id": "A group of nodes.",
        "Nodes": [
          "61aa6da7-0f40-4fbd-b600-05e0daf7e70c",
          "6f0d0008-b446-40aa-be41-75fb3b375f39",
          "00fdf5cf-f509-40d0-afbb-e83957b19cf3",
          "3ea93ee6-7419-4e6a-9748-25737de09ff4"
        ]
      }
    ]
  }
}
```

**Comment by gregmarr** [comment]
`Graph` is already the name of the object that is represented there.  It seems weird to have a `Graph` object within a `Graph` object.  What is the purpose of this proposed extra layer?

**Comment by gregmarr** [comment]
I could *maybe* see putting `Nodes` and `Connectors` inside a `Compute` object, as those are the parts needed by the compute component if you want to explicitly call out "compute" vs "view" data, but the `ID`, `Name`, and `Description` aren't actually needed by compute.

**Comment by mjkkirschner** [comment]
We've had a few discussion where we described the graph as being the compute object. I think it is easier for us to reason about the view being a separable entity if it's not inside the graph.

The dynamo file format would be composed of both these objects - compute graph, and view.

**Comment by gregmarr** [comment]
That sounds like where we were prior to the `workspace` to `graph` renaming, with the overall thing being a `workspace`, that contains a `graph` within it.  The only difference was that it wasn't an explicit sub-object.  I don't particularly want to re-open that discussion.

**Comment by radumg** [comment]
I'm sure this would've been discussed at length internally so apologies if i'm stating the obvious. 
All the below is based on my having to consume a few APIs and write clients recently. Doing so, I've noticed that APIs that make developers' lives really hard (and by extension users) tend to do either of 2 things in their responses/structures :

#### Response formats that do not match the class structures
Examples would include :
- the `graph` inside a `graph` inception case highlighted by @gregmarr [above](https://github.com/DynamoDS/Dynamo/issues/7747#issuecomment-303522989) 👍 
- the introduction of additional abstractions such as a `Compute` [wrapper](https://github.com/DynamoDS/Dynamo/issues/7747#issuecomment-303523767), which aren't mirrored in the classes
- responses that require clients to implement new `response` classes to map received data to actual class. wrapping the structure above in a `data` object should be fine though, as deserialisers tend to support selection of a root object.

#### Deviation from widely adopted conventions/standards
Whilst there is no one universally accepted standard that i know of for REST API response formats, there are a few specifications that seem to have achieved widespread adoption, such as :
- [JSON API](http://jsonapi.org/format/#document-structure)
- [Google JSON style guide](https://google.github.io/styleguide/jsoncstyleguide.xml#JSON_Structure_&_Reserved_Property_Names)
- [JSend](https://labs.omniti.com/labs/jsend)

As examples, all three of these have strictly specified structures for the root response objects which would point to wrapping all the DYN objects above not in a `graph` element but a `data` element. `Data` would then hold the `graph`, `view` & other objects, encapsulating all objects required in the DYN format for compute (*graph*) and application use (*view*) but not anything about the format itself (*meta*).

These specs could also address the idea of `Move versioning from Dynamo to the graph` from @ikeough 's first post. In the JSON API spec, this could be either in [meta](http://jsonapi.org/format/#document-meta) or the [jsonapi](http://jsonapi.org/format/#document-jsonapi-object) objects.

Hopefully you'll find these observations and suggestions to be useful. 🙂

**Comment by mjkkirschner** [comment]
@gregmarr - the `workspace` was renamed to `graph` in compute service - not in Dynamo. Because a graph is used strictly for compute - but a `workspace` has other properties, namely view data - and some other meta data. 

`graph` or `compute` is a subset of workspace model I think.

When calling a compute service - it makes sense to me to be able to grab strictly the object used for compute, without walking the dynamo file for required properties, and send this object to that service.

If I send a json object that has a view object - will a graph service reject my request?

**Comment by ikeough** [comment]
@radumg @mjkkirschner @QilongTang Thanks for all the suggestions. I believe we're conflating the `Graph` and what's in the .dyn file. This is completely understandable given that we've interwoven these conversations here and in the proposal above. What @QilongTang has shown is the .dyn file. 

Some further assertions:
- A `Workspace` should contain a `Graph`. Currently they are the same. The `Graph` class should be moved into a separate `Graph.cs` file.
- The `GraphViewModel` should also be a separate class in `WorkspaceViewModel`, with a corresponding `GraphViewModel.cs`.

- `Workspace` should contain 
- A `WorkspaceViewModel` should contain a `GraphViewModel`.
- The .dyn file should be expected to contain both `Graph` and `View` models.
- A 

Serialization will happen in two parts. First, the Workspace

The `View` has to be separate from the `Graph` because if we move the view-specific properties off graph models, there will need to be a separate deserialization step to deserialize the `GraphViewModel` model. This will happen when the `GraphViewModel` has been constructed completely. No `GraphViewModel` no deserialization of the `View`.


@radumg Thanks very much for the comments. I'm not sure I agree with your last assertion about wrapping everything in a `data` object. As initially proposed in this issue, the compute items (nodes, connectors, etc.) being at the "root", would allow direct deserialization of the json into a `Graph` model.

@mjkkirschner @QilongTang I agree with @gregmarr. I don't want to embed the graph in a `graph` object. However, your comments make me realize that we're conflating the graph schema and what's in the .dyn file. 

about the distinction between the .dyn file and what we are discussing here. The .dyn file can have two parts, one of which is the `Graph`, the other of which is the `View`.

**Comment by ramramps** [comment]
@ikeough Thanks for the clarification. So, the changes in Dynamo Core will reflect the above format? In that case,  is there any changes needed in CoGS to support this? Just to ensure we cover all these aspects in the tasks that Green team created for this and next sprint.

**Comment by mjkkirschner** [comment]
@ikeough - apologies if my comments were not clear, thanks for clarifying the crux of the issue. I agree with your comments above.

**Comment by mjkkirschner** [comment]
@ikeough @gregmarr @QilongTang @ramramps 
adding what I believe is the most up to date sample graph.
note graph has Uuid while all other models have Id
node Inputs / Outputs vs InputPorts OutputPorts is still under discussion...

```json
{
  "graph":{
    "Uuid": "f170f3ea-ae94-43c1-af0c-e2fa76ebc725",
    "Name":"My graph",
    "Description":"A compute graph.",
    "Nodes":[
      {
        "Id":"aNode",
        "ConcreteType":"Dynamo.Graph.Nodes.ZeroTouch.DSFunction, DynamoCore",
        "NodeType":"FunctionNode",
        "FunctionSignature":"Autodesk.DesignScript.Geometry.Point.ByCoordinates@double,double,double",
        "VariableInputPorts": true,
        "Inputs":[
          {
            "Id":"aPort",
            "UseLevels":false,
            "SelectedLevel":2,
            "KeepListStructure":false
          }
        ],
        "Outputs":[],
        "Replication":"auto"
      }
    ],
    "Connectors":[
      {
        "Id": "593c497e-08e3-485a-bf5d-1943d87a9c3b",
        "Start": "38727d94-86bd-4db8-b2a3-95baa1ac8a85",
        "End": "54c0f71a-524b-4443-9972-1e433a0dceb4"
      }
    ],
    "Dependencies":[
      "b7d9beae-14a0-485b-b259-27d0861eb165",
      "e1764a07-f1f4-4257-a672-9920cd7acc4a"
    ],
    "Bindings":[]
  }
  "View":{
    "X":0.0,
    "Y":1000.0,
    "Zoom":0.5,
    "NodeViews":[
      {
        "d65016b6-6845-4423-a831-d20cbf7d37d4":{
          "Name":"Point.ByCoordinates",
          "X":0.0,
          "Y":5.0,
          "IsVisible":"true",
          "IsUpstreamVisible":true
        }
      }
    ],
    "Cameras":[],
    "Annotations":[
      {
        "Id": "A group of nodes.",
        "Nodes": [
          "61aa6da7-0f40-4fbd-b600-05e0daf7e70c",
          "6f0d0008-b446-40aa-be41-75fb3b375f39",
          "00fdf5cf-f509-40d0-afbb-e83957b19cf3",
          "3ea93ee6-7419-4e6a-9748-25737de09ff4"
        ]
      }
    ]
  }
}
```

**Comment by gregmarr** [comment]
I thought we dropped the "graph" top level element.

**Comment by mjkkirschner** [comment]
Sorry we're still not the same page, but I (and I believe the green team) took the above comment by @ikeough - to mean that eventually there will be a graph class in Dynamo, which will serialize to `graph` and contain the workspace model serialization. Meanwhile View is a separate block.

This should have no large impact on any compute service, as the graph is what is sent for compute.

I think this is within the bounds of the task that @ikeough is currently working on and will defer to what he finds there.

**Comment by gregmarr** [comment]
I'm trying to track down the discussion that @ikeough and I had about this.  It was in DM on May 26th, but was about a more public document.  I haven't found it yet.

**Comment by gregmarr** [comment]
```
    "NodeViews":[
      {
        "d65016b6-6845-4423-a831-d20cbf7d37d4":{
          "Name": "Point.ByCoordinates",
          "X": 0.0,
          "Y": 5.0,
          "IsVisible": true,
          "IsUpstreamVisible": true
        }
      }
    ],
```
This doesn't work well in Swagger.  Can we do this instead:
```
    "NodeViews":[
      {
        "Uuid": "d65016b6-6845-4423-a831-d20cbf7d37d4",
        "Name": "Point.ByCoordinates",
        "X": 0.0,
        "Y": 5.0,
        "IsVisible": true,
        "IsUpstreamVisible": true
      }
    ],
```

**Comment by mjkkirschner** [comment]
How about? 
`Id` so it matches the Node serialization.
"NodeViews":[
      {
        "Id": "d65016b6-6845-4423-a831-d20cbf7d37d4",
        "Name": "Point.ByCoordinates",
        "X": 0.0,
        "Y": 5.0,
        "IsVisible": true,
        "IsUpstreamVisible": true
      }
    ],

**Comment by gregmarr** [comment]
@mjkkirschner  Makes sense. Can you also give me descriptions of `IsVisible` and `IsUpstreamVisible` for `NodeView`?  Should `Annotation` have a `Name` field?

**Comment by mjkkirschner** [comment]
* today `isVisible` reflects the preview state of geometry (is it showing or not)
* I *think*`IsUpstreamVisible`reflects if the node outputs its upstream nodes into render packages - this is only useful for watch 3d nodes, and I don't know if it works very well.... I think we could get rid of this from Dynamo in the future.
* I do not think `Annotation` has a name, though it has a text field.
* I proposed nodes just had ids(@pboyer initially suggested this I think)  and no one disagreed when I asked in my PR :) - I've added a set of tests that duplicate the serializationTests but with non guid ids for all models except the `workspaceModel`

**Comment by gregmarr** [comment]
I knew we agreed that non-uuid for ports, wasn't sure about nodes.

**Comment by mjkkirschner** [comment]
@gregmarr @ikeough 
* I'm not sure about the `isVisible` rename... no strong feelings - but that should be straight forward.
* larger question - does the view block belong in the schema at all?

**Comment by gregmarr** [comment]
>larger question - does the view block belong in the schema at all?

Absolutely.  The schema should contain exactly those things that every client needs to display the graph. 
 Otherwise, every client is going to be inventing their own method of storing it, and no one will be able to display anyone else's graph properly.

**Comment by QilongTang** [comment]
Latest schema:
{
  "Uuid": "772c4d87-44cd-42d2-9971-1c236b98e25f",
  "IsCustomNode": false,
  "Description": null,
  "Name": "Home",
  "ElementResolver": {
    "ResolutionMap": {}
  },
  "Nodes": [
    {
      "ConcreteType": "Dynamo.Graph.Nodes.CodeBlockNodeModel, DynamoCore",
      "NodeType": "CodeBlockNode",
      "Code": "x=1;",
      "Id": "26c5d862-ea26-496d-97f7-15c3e03e3e18",
      "Inputs": [],
      "Outputs": [
        {
          "Id": "7d61a73d-52fd-4aa0-894a-c8b9a628b44e",
          "Name": "",
          "Description": "x",
          "Level": 2,
          "UseLevels": false,
          "KeepListStructure": false
        }
      ],
      "Replication": "Disabled",
      "Description": "Allows for DesignScript code to be authored directly"
    }
  ],
  "Notes": [],
  "Connectors": [],
  "Dependencies": [],
  "Bindings": [],
  "View": {
    "Cameras": {
      "EyePosition": "-17,24,50",
      "UpDirection": "0,1,0",
      "LookDirection": "12,-13,-58",
      "Name": "Background Preview"
    },
    "NodeViews": [
      {
        "26c5d862-ea26-496d-97f7-15c3e03e3e18": {
          "Name": "Code Block",
          "X": 152.0,
          "Y": 89.0,
          "IsVisible": true,
          "IsUpstreamVisible": true
        }
      }
    ],
    "Annotations": [],
    "X": 0.0,
    "Y": 0.0,
    "Zoom": 1.0
  }
}

**Comment by mjkkirschner** [comment]
@QilongTang when "Notes" are serialized they should not have a `UUID` right? There is a task for serializing them as `Annotations` correct?

**Comment by ramramps** [comment]
That is on top of the backlog. Not sure when we will visit that task. 
The notes should live inside View block. Do we have a task for it? @QilongTang can you create one if there is not already?

**Comment by QilongTang** [comment]
@mjkkirschner Yes, specified in https://jira.autodesk.com/browse/QNTM-908, can we solve them together?

**Comment by ramramps** [comment]
Moving the notes to view block should be done first. It will help the deserilization. Then we can visit this task and do the serialization / deserialization as specified. Makes sense @QilongTang ?

**Comment by QilongTang** [comment]
https://jira.autodesk.com/browse/QNTM-923 and https://jira.autodesk.com/browse/QNTM-908 for moving `Notes` from graph block into View->Annotation block

**Comment by radumg** [comment]
Quick question on the schema above : will the `"Nodes": []` array also allow specification of an entire graph as an object, for use with graph orchestration when/if this ships ?

If there is an intention to enable this workflow later on (we absolutely should in my opinion, really need it in practical applications) , would this require the addition of another property at top-level ? Something like :
```
{
  "Uuid": "772c4d87-44cd-42d2-9971-1c236b98e25f",
  "IsCustomNode": false,
  **"IsEmbeddedGraph": true,**
  "Description": null,
  "Name": "Home",
  "ElementResolver": {
    "ResolutionMap": {}
  },
  "Nodes": [],
   "Notes": [],
  "Connectors": [],
  "Dependencies": [],
  "Bindings": [],
  "View": {}
}
```
Let me know if it's off-topic or for another thread.

**Comment by QilongTang** [comment]
To those still watching this thread, I am closing this since the team has finished the framework of Json serialization as part of Dynamo 2.0. If you have request or questions about schema, please open new issue or PR.
